### PR TITLE
Backend feature article

### LINKED ISSUE #266: Backend Add Article Page
Article is a text written by user about trading equipment. As we indicated in the Requirements Analysis:

1.1.2.5.1 User shall be able to share ideas about trading equipment as articles .

### PR DESCRIPTION
**Pull request by cemalaytekin**
closes #266

### CODE REVIEW

File: app/backend/controllers/article.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  }
+  
+  /*
+    Get method for articles.
+    It returns given article.
+  */
+  module.exports.getArticle = async (request, response) => {
+    let Article = request.models['Article']
+
+    article = await Article.findOne({ _id : request.params['id']});
```
Comment (gurkandemir) [inline app/backend/controllers/article.js]: we also return owners name and surname as we did in followers. you can checkout utils.js

File: app/backend/controllers/article.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        errmsg: "No such article."
+      })
+    }
+  }
+
+/*
+  Patch method for editing articles.
+  It saves article to database.
+*/
+module.exports.editArticle = async (request, response) => {
```
Comment (gurkandemir) [inline app/backend/controllers/article.js]: we will write it in more modular way

File: app/backend/controllers/article.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    let Article = request.models['Article']
+  
+    Article.deleteOne({ _id : request.params['id'], userId : request.session['user']._id }, (err, results) => {
+      if(err){
+        return response.status(404).send({
+          errmsg: "Failed."
+        })
+      }
+  
+      return response.status(404).send();
```
Comment (gurkandemir) [inline app/backend/controllers/article.js]: 404?

File: app/backend/controllers/profile.js [hunk 4]
```diff
@@ -213,3 +216,22 @@ module.exports.rejectRequest = async (request, response) => {
   })
 }
 
+/*
+  Get method for articles of user.
+*/
+module.exports.getArticles = async (request, response) => {
```
Comment (gurkandemir) [inline app/backend/controllers/profile.js]: do we call this method?

File: app/backend/controllers/article.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  let Article = request.models['Article']
+  const userId = request.session['user']._id
+  const title = request.body["title"];
+  const text = request.body["text"];
+  const articleId = request.body['articleId'];
+
+  article = await Article.findOne({ _id : articleId});
+  if(article){
+    try{
+      if(!title && !text){ // no change
```
Comment (gurkandemir) [inline app/backend/controllers/article.js]: you do not have to check whether title, text exist or not. client will request you in correct format, you only have to update the article with given id.

even if title did not change, client must send you existing title in the request body.

Comment (gurkandemir) [reply app/backend/controllers/article.js]: you can think updating as semi-creating. the only difference is that you have to add new row to database with creating new _id. but in updating you change only existing row with given _id.

Comment (gurkandemir) [inline app/backend/controllers/article.js]: you do not have to check whether title and text exist or not. client must request you in true format.

even if title did not change, client must add existing title in to request body.

you can think updating as semi-creating. the only difference is that in creating you have to add new row to database with creating new _id.

but in updating you change only existing row with given _id.

File: app/backend/controllers/profile.js (hunk 4 above)
Comment (cemalaytekin) [reply app/backend/controllers/profile.js]: At some point, if someone needs a specific article, he/she can use this method.

File: app/backend/controllers/article.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            return response.status(400).send(error);
+          });
+          return response.status(204).send();
+        }).catch(error => {
+          return response.status(400).send(error);
+        });
+  
+      } else{
+        // User already rated for that article. Change the rate value and update the database
+        await ArticleUser.updateOne({_id:row._id},{ rate: value})
```
Comment (bahadirh) [inline app/backend/controllers/article.js]: Since that line onward we've used promises, using await here may produce unexpected results.

Comment (bahadirh) [reply app/backend/controllers/article.js]: another instance of that in three lines below

File: app/backend/controllers/profile.js (hunk 4 above)
Comment (bahadirh) [inline app/backend/controllers/profile.js]: The endpoint may be changed to singular and console logs can be removed too.

Comment (cemalaytekin) [reply app/backend/controllers/profile.js]: Actually we will never call this function. So I will delete it.

**Pull request merged**